packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -568,3 +569,143 @@ class _TextSelectionHandleOverlayState extends State<_TextSelectionHandleOverlay
     return null;
   }
 }
+
+/// A gesture detector to detect concurrent gesture events for a text field.
+///
+/// Unlike a plain [GestureDetector] where a single gesture wins the gesture
+/// arena, text fields supports chains of concurrent events.
+///
+/// See also:
+///
+///  * [TextField], a Material text field which uses this gesture detector.
+///  * [CupertinoTextField], a Cupertino text field which uses this gesture
+///    detector.
+class TextSelectionGestureDetector extends StatefulWidget {
+  /// Create a [TextSelectionGestureDetector].
+  ///
+  /// Multiple callbacks can be called from the same sequence of gesture.
+  /// The [child] parameter must not be null.
+  const TextSelectionGestureDetector({
+    this.onTapDown,
+    this.onSingleTapUp,
+    this.onSingleTapCancel,
+    this.onSingleLongTapDown,
+    this.onDoubleTapDown,
+    @required this.child,
+  }) : assert(child != null);
+
+  /// Called each distinct time a tap was held momentarily or on a short tap.
+  final GestureTapDownCallback onTapDown;
+
+  /// Called for each distinct short tap. Also called on the first tap of a
+  /// double tap but not on the second tap.
+  final GestureTapUpCallback onSingleTapUp;
+
+  /// Called for each tap down that does not become a short tap such as a long
```

Comment:
I think it would be better to explain when this is called. I assume it's called because a tap down occurred - and onTapDown was called, and the pointer only moved a little and was held down long enough to qualify as a "long tap down".

Also: there's no shame in mentioning kDoubleTapTimeout and kDoubleTapSlop

Reply:
Added some more details. It's not really related to those 2 constants.